Userland/Libraries/LibPDF/Renderer.cpp

```diff
@@ -307,32 +308,101 @@ RENDERER_HANDLER(path_append_rect)
     return {};
 }
 
-PDFErrorOr<void> Renderer::add_clip_path(Gfx::WindingRule)
+PDFErrorOr<void> Renderer::prepare_clipped_bitmap_painter()
+{
+    if (!m_clipped_bitmap)
+        m_clipped_bitmap = TRY(Gfx::Bitmap::create(Gfx::BitmapFormat::BGRA8888, m_bitmap->size()));
+
+    m_clipped_bitmap_painter = make<Gfx::Painter>(*m_clipped_bitmap);
+    m_clipped_bitmap_anti_aliasing_painter = make<Gfx::AntiAliasingPainter>(*m_clipped_bitmap_painter);
+
+    auto r = state().clipping_state.clip_bounding_box;
+    m_clipped_bitmap_painter->add_clip_rect(r);
+
+    // Need to copy what's currently on the page so that transparent things in the clipped area are blended on top of the right things.
+    m_clipped_bitmap_painter->blit(r.location(), *m_bitmap, r);
+
+    return {};
+}
+
+static void apply_clip(Gfx::Bitmap& bitmap, Gfx::IntRect bitmap_rect, Gfx::Bitmap const& clip_alpha, Gfx::IntRect clip_rect)
+{
+    VERIFY(bitmap.size() == bitmap_rect.size() || (bitmap_rect.is_empty() && bitmap.size() == Gfx::IntSize(1, 1)));
+    VERIFY(clip_alpha.size() == clip_rect.size() || (clip_rect.is_empty() && clip_alpha.size() == Gfx::IntSize(1, 1)));
+    auto rect = bitmap_rect.intersected(clip_rect);
+
+    for (int y = rect.top(); y < rect.bottom(); ++y) {
+        for (int x = rect.left(); x < rect.right(); ++x) {
+            auto pixel = bitmap.get_pixel(Gfx::IntPoint { x, y } - bitmap_rect.location());
+            auto alpha = clip_alpha.get_pixel(Gfx::IntPoint { x, y } - clip_rect.location());
+            pixel.set_alpha(pixel.alpha() * alpha.alpha() / 255);
+            bitmap.set_pixel(Gfx::IntPoint { x, y } - bitmap_rect.location(), pixel);
```

Comment:
This seems a bit repetitive imo

Reply:
How would you write it?

Reply:
Maybe pull out a `Gfx::IntPoint origin { x, y };` or what the appropriate name is

Reply:
Oh, that's what you mean :) Sure, done!